src/Compilers/CSharp/Test/Emit3/OverloadResolutionPriorityTests.cs

```diff
@@ -0,0 +1,1591 @@
+    public void AmbiguityWithinPriority(int priority)
+    {
+        var source = $$"""
+            using System.Runtime.CompilerServices;
+
+            I3 i3 = null;
+            C.M(i3);
+
+            interface I1 {}
+            interface I2 {}
+            interface I3 : I1, I2 {}
+
+            class C
+            {
+                [OverloadResolutionPriority({{priority}})]
+                public static void M(I1 x) => System.Console.WriteLine(1);
+
+                [OverloadResolutionPriority({{priority}})]
+                public static void M(I2 x) => throw null;
+            }
+            """;
+
+        CreateCompilation([source, OverloadResolutionPriorityAttributeDefinition]).VerifyDiagnostics(
+            // (4,3): error CS0121: The call is ambiguous between the following methods or properties: 'C.M(I1)' and 'C.M(I2)'
+            // C.M(i3);
+            Diagnostic(ErrorCode.ERR_AmbigCall, "M").WithArguments("C.M(I1)", "C.M(I2)").WithLocation(4, 3)
+        );
+    }
+
+    [Theory, CombinatorialData]
+    public void MethodDiscoveryStopsAtFirstApplicableMethod(bool useMetadataReference)
+    {
+        var @base = """
+            using System.Runtime.CompilerServices;
+
+            public interface I1 {}
+            public interface I2 {}
+            public interface I3 : I1, I2 {}
+
+            public class Base
+            {
+                [OverloadResolutionPriority(1)]
+                public static void M(I2 x) => throw null;
+            }
+            """;
+
+        var derived = """
+            public class Derived : Base
+            {
+                public static void M(I1 x) => System.Console.WriteLine(1);
+            }
+            """;
+
+        var executable = """
+            I3 i3 = null;
+            Derived.M(i3);
+            """;
+        CompileAndVerify([executable, @base, derived, OverloadResolutionPriorityAttributeDefinition], expectedOutput: "1").VerifyDiagnostics();
+
+        var baseComp = CreateCompilation([@base, OverloadResolutionPriorityAttributeDefinition]);
+        var baseReference = useMetadataReference ? baseComp.ToMetadataReference() : baseComp.EmitToImageReference();
+
+        var derivedComp = CreateCompilation(derived, references: [baseReference]);
+        var derivedReference = useMetadataReference ? derivedComp.ToMetadataReference() : derivedComp.EmitToImageReference();
+
+        CompileAndVerify(executable, references: [baseReference, derivedReference], expectedOutput: "1").VerifyDiagnostics();
+    }
+
+    [Theory, CombinatorialData]
+    public void MethodDiscoveryStopsAtFirstApplicableIndexer(bool useMetadataReference)
+    {
+        var @base = """
+            using System.Runtime.CompilerServices;
+
+            public interface I1 {}
+            public interface I2 {}
+            public interface I3 : I1, I2 {}
+
+            public class Base
+            {
+                [OverloadResolutionPriority(1)]
+                public int this[I2 x]
```

Comment:
>public int this[I2 x] [](http://example.com/codeflow?start=16&length=21)

Isn't this indexer dropped from the candidate set even before we look at overload resolution priority?

Reply:
Yes, that's the point of this test.